rfcs/20230420-kernel-cache/README.md

```diff
@@ -0,0 +1,292 @@
+`ONEDNN_ENABLE_KERNEL_CACHE`.
+
+##### Options Comparison
+
+Both options assume that there will be a reasonable default capacity for the
+kernel cache. Also, both options allow users to control the caches capacity
+therefore, the only difference between the options is in the flexibility.
+The second option is in fact the first one with advanced controlling
+capabilities.
+
+Given that the first option will address the problem with dynamic shapes
+related creation overhead the proposal is to start with the it and add the
+API from the option two if we see any need in it.
+
+### Option 2: Extending Runtime Dimensions Support
+
+This option suggests extending the existing runtime dimensions support for
+relevant primitives.
+
+Pros:
+* Using the existing mechanism, the primitive cache will automatically work
+for primitives created with runtime dimensions
+* Unlike the kernel cache, with this approach the primitive and primitive descriptor
+is not re-created therefore there is no related overhead
+
+Cons:
+* The users have to modify their code to enable runtime dimensions
+* Some primitives will require to generate multiple kernels so that the
+primitives can be used with runtime dimensions. It may not be always feasible to
+generate a reasonable number of kernels
+* For some kernels it may not be possible to stay performant and support runtime
+dimensions for all dimensions. This means that if user uses incorrect combination
+of runtime dimensions they may get suboptimal performance
+    * There is an option to document all combinations that will not introduce
+    performance regressions however this would be essentially exposing
+    implementation details to the users. This will require us to make sure that
+    the documented combinations are valid for all hardware and implementations
+    which will have maintenance cost
+
+
+### Conclusion
+
+The runtime dimensions feature eliminates most creation overhead (primitive descriptor,
+primitive, kernel). However, the primitives may have to generate multiple kernels to
+support the feature. And if there are too many (or even indefinite amount of) kernels
+to generate then the runtime dimensions cannot be supported or can be supported partially
+which would mean the supported combinations will have to be documented (i.e., exposing
+implementation details).
+
+The kernel cache feature eliminates only kernel creation overhead that is usually the
+most significant one. However, it will work for all primitives whenever the kernels can be
+re-used. With this feature there is no need to generate all possible kernels, only those
+that are used in the workload will be generated and re-used.
+
+As we can see there is not an ideal solution to the original problem. The conclusion is
+that the runtime dimensions feature work better for optimizations very specific cases
+but provide does the best job at eliminating any creation overhead. While the kernel
```

Comment:
It's not clear what is meant by `but provide does the best job...`. Can you please reword it?